### PR TITLE
Fix check for form parameters that don't need complex binding in RDG

### LINKED ISSUE #55840: [9.0-preview.5] Endpoint with nullable parameter no longer generates code with Request Delegate Generator
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Using version `9.0.100-preview.5.24272.3` of the .NET SDK from daily builds, an endpoint associated with a method that has a `string[]?` parameter is no longer having an endpoint generated an produces a RDG003 warning.

```console
Error: /home/runner/work/adventofcode/adventofcode/src/AdventOfCode.Site/Program.cs(135,1): error RDG003: Unable to statically resolve parameter named 'arguments' for endpoint. Compile-time endpoint generation will skip this endpoint and the endpoint will be generated at runtime. For more information, please see https://aka.ms/aspnet/rdg-known-issues (https://learn.microsoft.com/aspnet/core/fundamentals/aot/request-delegate-generator/diagnostics/RDG003) [/home/runner/work/adventofcode/adventofcode/src/AdventOfCode.Site/AdventOfCode.Site.csproj]
```

The error is not generated with .NET 8 or .NET 9 preview 4.

The signature of the endpoint is the following ([code](https://github.com/martincostello/adventofcode/blob/fe22226c9428f995e396a9ee5f81b8c7ae6dbd7d/src/AdventOfCode.Site/PuzzlesApi.cs#L55-L64)):

```chsarp
internal static async Task<IResult> SolvePuzzleAsync(
    int year,
    int day,
    HttpRequest request,
    [FromForm] string[]? arguments,
    IFormFile? resource,
    PuzzleFactory factory,
    TimeProvider timeProvider,
    ILogger<Puzzle> logger,
    CancellationToken cancellationToken)
```

### Expected Behavior

Request Delegate Generator successfully generates code for this endpoint.

### Steps To Reproduce

1. Clone https://github.com/martincostello/adventofcode/pull/1663/commits/fe22226c9428f995e396a9ee5f81b8c7ae6dbd7d#diff-f905acb7a057826fd183f2558445347c59e1cf0304dc0499d9f5e0ae4a523ebe
2. Run `build.ps1` in the root of the repository.

### Exceptions (if any)

_No response_

### .NET Version

9.0.100-preview.5.24272.3

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/55840.

https://github.com/dotnet/aspnetcore/blob/83573c72d12e1c2018805bc5d461d730bcb9dcc4/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs#L110-L119

The condition added in the code referenced above needs an additional `(IsArray && ElementType.SpecialType == SpecialType.System_String)` check since we don't don't generate a parsability method for string types out of TryGetParsability.

### CODE REVIEW

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 1]
```diff
@@ -111,6 +111,7 @@ private void ProcessEndpointParameterSource(Endpoint endpoint, ISymbol symbol, I
             // and codegen. Emit a diagnostic when these are encountered to avoid producing buggy code.
             else if (!(SymbolEqualityComparer.Default.Equals(Type, wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_Primitives_StringValues))
                     || Type.SpecialType == SpecialType.System_String
+                    || (IsArray && ElementType.SpecialType == SpecialType.System_String)
                     || TryGetParsability(Type, wellKnownTypes, out var _)
                     || (IsArray && TryGetParsability(ElementType, wellKnownTypes, out var _))))
```
Comment (martincostello) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:116]: I think I understand now why I thought this was an intentional change to emit the diagnostic when I looked at fixing this myself - I misread the braces and thought it was `!a or b or c or d` rather than `!(a or b or c or d)`.

Is it worth adjusting the alignment slightly to improve the readability? Subjective opinion, might just be that **I** misread it and it's clear to others.

```suggestion
            else if (!(SymbolEqualityComparer.Default.Equals(Type, wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_Primitives_StringValues))
                       || Type.SpecialType == SpecialType.System_String
                       || (IsArray && ElementType.SpecialType == SpecialType.System_String)
                       || TryGetParsability(Type, wellKnownTypes, out var _)
                       || (IsArray && TryGetParsability(ElementType, wellKnownTypes, out var _))))
```

Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:116]: Good feedback -- I'll do you one better as well and move this to a seperate method.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 2]
```diff
@@ -109,10 +109,7 @@ private void ProcessEndpointParameterSource(Endpoint endpoint, ISymbol symbol, I
             }
             // Complex form binding is only supported in RDF because it uses shared source with Blazor that requires dynamic analysis
             // and codegen. Emit a diagnostic when these are encountered to avoid producing buggy code.
-            else if (!(SymbolEqualityComparer.Default.Equals(Type, wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_Primitives_StringValues))
-                    || Type.SpecialType == SpecialType.System_String
-                    || TryGetParsability(Type, wellKnownTypes, out var _)
-                    || (IsArray && TryGetParsability(ElementType, wellKnownTypes, out var _))))
+            else if (RequiresComplexFormBinding(wellKnownTypes))
```
Comment (halter73) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:112]:
```suggestion
            else if (!HasSimpleBinding(wellKnownTypes))
```

The `!` inside the `RequiresComplexFormBinding` is still hard to parse with the nested parentheses. I think moving it outside the method would make it clearer.

**Review by halter73 — APPROVED** [review]

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 3]
```diff
@@ -109,10 +109,7 @@ private void ProcessEndpointParameterSource(Endpoint endpoint, ISymbol symbol, I
             }
             // Complex form binding is only supported in RDF because it uses shared source with Blazor that requires dynamic analysis
             // and codegen. Emit a diagnostic when these are encountered to avoid producing buggy code.
```
Comment (halter73) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:111]:
```suggestion
            // Minimal APIs shares the same implementation that Blazor uses for complex form binding which uses reflection.
            // We could implement a source gen layer for form binding but the plan has been to do that in conjunction with Blazor,
            // so we're not solving the same problem twice in different ways.
```

I couldn't follow this. Can we just rewrite it to match [your PR comment](https://github.com/dotnet/aspnetcore/pull/55465#discussion_r1586881468) on the PR that added it?

Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs]: Ah, good note. After re-reading this, I realize it's not particularly helpful for understanding the reasoning. 😅 

Tweaked the verbiage here a bit.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs]: Fixed.
